diff --git a/applications/CMakeLists.txt b/applications/CMakeLists.txt
index 178fc53..19a426c 100644
--- a/applications/CMakeLists.txt
+++ b/applications/CMakeLists.txt
@@ -22,3 +22,7 @@ if (CMAKE_CXX_COMPILER_ID STREQUAL "ARMClang")
     # Only armclang supported for now
     add_subdirectory(trustzone_inference)
 endif()
+
+add_subdirectory(hello_tflm)
+add_subdirectory(mobilenet_v2)
+add_subdirectory(person_detection)
